File: DataSets/DataSet-New/LeetCodeTasks - Develop/LeetCodeTasks/3068_find-the-maximum-sum-of-node-values/solutions/cpp/solution.cpp
#include <vector>
#include <algorithm>

using namespace std;

class Solution {
public:
    long long maximumValueSum(vector<int>& nums, int k, vector<vector<int>>& edges) {
        int n = nums.size();
        vector<vector<int>> adj(n);
        for (const auto& edge : edges) {
            adj[edge[0]].push_back(edge[1]);
            adj[edge[1]].push_back(edge[0]);
        }

        vector<int> xor_change(n, 0);  // Tracks the cumulative XOR to apply at each node
        long long max_sum = 0;
        
        function<void(int, int)> dfs = [&](int node, int parent) {
            long long original = nums[node];
            long long toggled = nums[node] ^ k;
            max_sum += max(original, toggled);

            for (int neighbor : adj[node]) {
                if (neighbor == parent) continue;
                // The parent's toggled value affects whether we toggle this node or not
                nums[neighbor] ^= xor_change[node];  // Apply cumulative XOR changes
                xor_change[neighbor] ^= xor_change[node];  // Propagate the change forward
                
                // Consider toggling at this node
                if (toggled > original) {
                    xor_change[neighbor] ^= k;  // This means we toggle at this node
                }

                dfs(neighbor, node);
            }
        };

        dfs(0, -1);
        
        return max_sum;
    }
};

File: DataSets/DataSet-New/LeetCodeTasks - Develop/LeetCodeTasks/3068_find-the-maximum-sum-of-node-values/solutions/java/solution.java
import java.util.ArrayList;
import java.util.List;

class Solution {
    public long maximumValueSum(int[] nums, int k, int[][] edges) {
        int n = nums.length;
        List<List<Integer>> graph = new ArrayList<>();
        for (int i = 0; i < n; i++) {
            graph.add(new ArrayList<>());
        }
        for (int[] edge : edges) {
            int u = edge[0], v = edge[1];
            graph.get(u).add(v);
            graph.get(v).add(u);
        }
        
        long[] maxSum = new long[1];
        boolean[] visited = new boolean[n];
        
        dfs(0, -1, nums, k, graph, visited, maxSum);
        
        return maxSum[0];
    }
    
    private void dfs(int node, int parent, int[] nums, int k, List<List<Integer>> graph, boolean[] visited, long[] maxSum) {
        visited[node] = true;
        
        long include = nums[node]; // initial value of node
        long exclude = nums[node] ^ k; // value if XOR applied
        
        for (int neighbor : graph.get(node)) {
            if (neighbor == parent) continue;
            
            dfs(neighbor, node, nums, k, graph, visited, maxSum);
            
            include += Math.max(nums[neighbor], nums[neighbor] ^ k); // max without changing the subtree of neighbor
            exclude += Math.max(nums[neighbor], nums[neighbor] ^ k); // max if change is made on current node
        }
        
        nums[node] = (int) Math.max(include, exclude); // best achievable configuration for this subtree
        maxSum[0] = Math.max(maxSum[0], Math.max(include, exclude));
    }
}
